### DISCUSSION TITLE
Safer alternatives to eval() in config and deserialisation paths

### DISCUSSION THREAD

**Discussion by falconsmilie** [MannLabs/scPortrait#393] [Ideas]
Hi.

The project uses `eval()` in a few config and file deserialisation paths.

This is risky because `eval()` executes Python code. If a config value or generated intermediate file is malformed or tampered with, it could execute arbitrary code in the user’s environment.

Even if these inputs are usually trusted, avoiding `eval()` would make the code safer, and easier to validate, test, and maintain.
___

A concrete example is the config-driven model class loading path.

If the code evaluates a config value like:

```python
eval(self.config["model_class"])
```

then the config does not only describe data; it can execute Python expressions.

For example, a config value such as:

```python
model_class: "__import__('pathlib').Path('eval_poc.txt').write_text('eval executed')"

```

would create a local file as soon as `eval(...)` runs. The pipeline may fail afterwards because the expression does not return a valid model class, but the side effect has already happened.

It is also possible for an expression to perform a side effect and still return a valid-looking value, for example:

```python
model_class: "(__import__('pathlib').Path('eval_poc.txt').write_text('eval executed'), SomeModelClass)[1]"
```

A similar issue exists for eval-based deserialisation of intermediate files. A line in a persisted tile/window file could execute code and then return a valid-looking slice object.

A safer approach may be to replace the `eval()` usages with explicit registries for known classes/transforms and structured serialisation for persisted slice/window data.

**Comment by falconsmilie** [comment]
@sophiamaedler  This is the biggest engineering issue i found in scPortrait package. It is both a security risk and a maintainability risk.

Especially if it evaluates values from config files or generated intermediate files, it means a config can potentially execute arbitrary Python.

It may be a good idea to treat this as the next meaningful ticket/PR after the [HDF5 fixes](https://github.com/MannLabs/scPortrait/pull/392).

Good PR scope:

Replace eval() for model class lookup with an explicit registry.
Replace eval() for transforms with an explicit registry or safe parser.
Replace eval() for persisted slice/window data with JSON parsing.

But I would not do all eval fixes in one giant PR unless the changes are tiny. The better split is:

remove eval() from config-driven model/transform loading
remove eval() from mask/window/slice deserialisation

That keeps review easier.

Not every `.eval()` in the repo is bad. These are fine:

`model.eval()`

That is standard PyTorch evaluation mode.

The risky calls are these:

```
eval(self.config["model_class"])
eval(model_class)
eval(self.config["transforms"])
[eval(line) for line in f.readlines()]
[eval(line.strip()) for line in f.readlines()]
```

i'm happy to help out if you're looking to improve this repo architecturally. Within this project I have a list of 20 or so things which need some attention. Are you interested? My goal at the moment is to find open source repos which are serious scientific python projects, which need a little help with the software engineering side of things.

**Comment by falconsmilie** [comment]
@sophiamaedler It could be something simple to begin with:

### Model class registry

Instead of:
```
self.define_model_class(eval(self.config["model_class"]))
```
use a registry:
```
MODEL_CLASS_REGISTRY = {
    "MultilabelSupervisedModel": MultilabelSupervisedModel,
}
```
Then:
```
model_class_name = self.config["model_class"]
try:
    model_class = MODEL_CLASS_REGISTRY[model_class_name]
except KeyError as exc:
    raise ValueError(
        f"Unknown model_class '{model_class_name}'. "
        f"Available options: {sorted(MODEL_CLASS_REGISTRY)}"
    ) from exc

self.define_model_class(model_class)
```
That is safer and clearer.

If custom model classes are needed, they should be passed programmatically through `define_model_class(...)`, not executed from config.

### Transform config

This is trickier because users may currently configure something like:
```
transforms.Compose([...])
```
through a string. Replacing that directly with a registry is still possible, but it requires choosing a config shape.

A safer future config shape could be:
```
transforms:
  - name: Resize
    args:
      size: [224, 224]
  - name: Normalize
    args:
      mean: [0.5, 0.5, 0.5]
      std: [0.5, 0.5, 0.5]
```
Then a registry:
```
TRANSFORM_REGISTRY = {
    "Resize": transforms.Resize,
    "CenterCrop": transforms.CenterCrop,
    "Normalize": transforms.Normalize,
    "ToTensor": transforms.ToTensor,
}
```
The loader builds `transforms.Compose(...)` from known safe transform names and arguments.

This is more work than the model-class fix, so it may be worth doing in two steps:

first remove eval() from model_class
then handle transforms

There is also some persisted slice/window eval() which could be replaced with JSON.

**Comment by sophiamaedler** [comment]
absolutely agree we should implement this. Has been on my backlog list for awhile now (at least for the persisted slice/window eval). Happy to approve any PRs related to this :) 

For the transforms: rational for implementing the way it currently is was to allow maximal compatibility without need for additional dev work. But I guess this comes at a safety tradeoff. I would suggest we start with the JSON serialization, then the model registry and then maybe revist how we can handle the transform. Thanks again for your input and effort! This is much appreciated.
